### PR TITLE
feat: Packaged Executable

### PR DESCRIPTION
**Pull request by bdferris-v2**
**Summary:**

Introduces functionality to package the GTFS Validator as an installable application, including a bundled JRE, to make it easier for users to run the validator.  As discussed in issue #1112, https://bit.ly/gtfs-validator-packaged-exe, and #1124.

Note: This is an initial draft review to get agreement on package structure before additional follow-up work.

**Expected behavior:** 

Prerequisites:
* A Windows or Mac OS build machine (I haven't tested on Linux, but it should theoretically work there as well)
* A Java SDK that includes `jlink` and `jpackage`
* If building on Windows, have https://wixtoolset.org/ installed on your path for Windows Installer support

To package:
* run `./gradlew jpackage`

Expected:
* You should find a "GTFS Validator" application and installer in the `app/pkg/build/jpackage` directory.

The current application is just a simple wrapper around the existing CLI.  On Windows, you can run the validator by dragging a GTFS zip file onto the application shortcut.  On Mac OS, you still have to run it from the command-line while we wait on some simple GUI support (e.g. `GTFS\ Validator.app/Contents/MacOS/GTFS\ Validator path/to/your/gtfs.zip`).

Future work will include:
* refactor the `main` sub-project to separate table+validation code from the cli code
* adding a simple UI for selecting a GTFS feed, output directory, configuring the validator, etc

### CODE REVIEW

**Comment by isabelle-dr** [comment]
@bdferris FYI, we recently implemented the [acceptance tests](https://github.com/MobilityData/gtfs-validator/blob/master/docs/ACCEPTANCE_TESTS.md) that runs this validator on all the datasets we have on the Mobility Database and flags if we had previously valid datasets that become invalid with the PR. 
It is part of our PR process to run these tests before merging any PR, but they currently fail if a PR is opened from a forked repo (see Issue #1094).
We haven't had a chance to have a look into it yet, but this is something that we will need to solve so we can run them on this PR.

cc @asvechnikov2

**Comment by barbeau** [comment]
I believe I have a workaround for #1094 that I'll comment on shortly.

**Comment by barbeau** [comment]
After https://github.com/MobilityData/gtfs-validator/pull/1102 is merged and the main branch is merged into this PR, then the acceptance tests should pass.

**Comment by bdferris-v2** [comment]
I think there might be some other test failures unrelated to issue #1094, as discussed in issue #1129?  Curious if you all have thoughts on that.

File: app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java
Comment (barbeau) [inline app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java:20]: If we're going to merge this PR as a first cut, should this TODO move to a GitHub issue?

**Review by barbeau — COMMENTED** [review]
Thanks @bdferris-v2, this looks good!

Some thoughts in-line.

**Comment by barbeau** [comment]
Also, looks like the Javadoc generation GitHub Action is failing:
https://github.com/MobilityData/gtfs-validator/runs/6276946492?check_suite_focus=true

```
....
  (package javax.annotation is declared in the unnamed module, but module org.mobilitydata.gtfsvalidator.app.pkg does not read it)
/home/runner/work/gtfs-validator/gtfs-validator/core/src/main/java/org/mobilitydata/gtfsvalidator/notice/RuntimeExceptionInLoaderError.java:3: error: package com.google.common.base is not visible
import com.google.common.base.Strings;
                        ^
  (package com.google.common.base is declared in module com.google.common, but module org.mobilitydata.gtfsvalidator.app.pkg does not read it)
100 errors

> Task :aggregateJavadocs FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':aggregateJavadocs'.
> Javadoc generation failed. Generated Javadoc options file (useful for troubleshooting): '/home/runner/work/gtfs-validator/gtfs-validator/build/tmp/aggregateJavadocs/javadoc.options'
```

File: app/pkg/build.gradle
Comment (bdferris-v2) [reply app/pkg/build.gradle:2]: What's the project policy on copyright headers?  I noticed it seems to be hit or miss across the codebase.  I'm wondering if I should go update all my files.  Also, I'm a Google employee but I've also signed the CLA, so I guess the copyright goes to MobilityData, yeah?

File: app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java
Comment (bdferris-v2) [reply app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java:66]: Macro comment here: I consider my approach here of calling the cli.Main with artificially constructed command-line arguments to be kind of a hack.  In a follow-up PR, I'd like to refactor this to call validator classes directly, but that kind of gets into the :main project refactor.

I think I will add a TODO for that with a link back to an issue?

File: app/pkg/build.gradle
Comment (barbeau) [reply app/pkg/build.gradle:2]: To my knowledge, past policy has been contributor retains copyright (Google has handled contributions this way in the past) and updates copyright header to include their organization and updates the date range to reflect copyright dates.

So if this contains copy of code from the existing project it would be:

`Copyright 2020-2022 Google LLC, MobilityData IO`

If it's brand new code without any existing project code it would be:

`Copyright 2022 Google LLC`

Comment (barbeau) [reply app/pkg/build.gradle:2]: Oh, and CLA contains provisions that allow MobilityData irrevocable re-use of code as part of this project.

File: app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java
Comment (barbeau) [reply app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java:66]: Sounds good! 👍 

We'll need that anyway for the roadmap of invoking the static validator from the RT validator.

**Comment by bdferris-v2** [comment]
@barbeau the aggregate javadoc issue is actually slightly tricky.  Javadoc generation appears to get tripped up for the :app:pkg sub-project because of the previously discussed tricks I used to get Java Modules working (shadow jar wrapped as a single module).  Specifically, Javadoc appears to be adding the the shadow jar its classpath along with all the other project dependencies, which is causing a bunch of duplicate class warnings.

There seem to be two options for fixing this:
1) We try to fully modularize the entire project.  I'm wary of going down this road for now.
2) We try to exclude the :app:pkg project from Javadoc generation.  It only has a single dummy class anyway, so we're not missing much there.

The trick with option # 2 is that we are using the `nebula-aggregate-javadocs` plugin from Netflix for project-wide javadoc aggregation.  This plugin doesn't appear to have a way of excluding a project (I've looked at the [source](https://github.com/nebula-plugins/gradle-aggregate-javadocs-plugin/blob/master/src/main/groovy/nebula/plugin/javadoc/NebulaAggregateJavadocPlugin.groovy)).  This plugin also hasn't been updated in five years?

As an alternative, there are newer Gradle plugins that support javadoc aggregation.  Specifically, I just tried the `io.freefair.aggregate-javadoc-jar`, which is currently being actively maintained.  Per the [source](https://github.com/freefair/gradle-plugins/blob/master/maven-plugin/src/main/java/io/freefair/gradle/plugins/maven/javadoc/AggregateJavadocPlugin.java), it's possible to exclude particular projects with a "javadoc { enabled = false }" clause in a sub-project.  I've verified this behavior on own project.  It seems to generate aggregated javadoc in build/docs/javadoc/ like the old plugin, though the target name has changed:

aggregateJavadocs => aggregateJavadoc (the s is dropped)

So I'd need to update the workflow as well.

I wanted to get a thumbs up from someone before I swapped out the plugin, in case anyone has any concerns.

**Comment by barbeau** [comment]
I noticed the Netflix plugin hadn't been updated in a while too, so I'm 👍 for changing to io.freefair.aggregate-javadoc-jar.

File: app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java
Comment (maximearmstrong) [reply app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java:70]: I also think that the directory of the input GTFS zip file may be a better option. 

Although, maybe I'm missing a point here. Why wouldn't the be organized in a single directory if the GTFS files aren't? Do you mean if there is multiple version of the GTFS zip file in the same directory rather than in subdirectories?

Comment (bdferris-v2) [reply app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java:70]: I think it's worth noting that I'd still like to support validation against a URL, so we might still need to find a directory for report output independent of the feed location.

I can think of a couple of additional options:
1) We create an explicitly named directory off the user's home directory to store reports.  You'll notice that I also added support for diagnostic logging to a $HOME/gtfs-validation.log file, so could be worth consolidating all artifacts in the same place?
2) We ask the user to pick an output directory the first time (or every time) that validation runs.  Or we at least remember their previous choice and pop-up a dialog with defaults and ask them to click "Validate" before things kick off.

**Comment by maximearmstrong** [comment]
I tested `./gradlew aggregateJavadoc` and it works well with `io.freefair.aggregate-javadoc-jar`. Thank you for this @bdferris-v2 ! 
I was also able to run `./gradlew jpackage` successfully, but I'm on MacOs so I can't test the application behavior for now. I'll be back with my full code review.

File: app/pkg/build.gradle
Comment (bdferris-v2) [reply app/pkg/build.gradle]: Addressed in 8f26673749679ad2efd01fc48fad89248c15a794.

File: app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java
Comment (bdferris-v2) [reply app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java:90]: It's a little tricky, because `cli.Main` appears to write both `report.json` and `system_errors.json`, regardless of the content of either.  So file existence isn't enough to know if the validator encountered system errors.  While I suppose I could attempt to open each json file to see if it had content, I think this is better addressed by waiting for the refactoring in #1135, which would allow us to examine the NoticeContainer directly.

Which is to say, consider this tracked by #1135.

File: app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java
Comment (bdferris-v2) [reply app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java:32]: Updated ARCHITECTURE.md in 398ed25fce7e144b3fe659f2f54c2c63250b819e.

File: app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java
Comment (bdferris-v2) [reply app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java:82]: I added a todo in 2251f9435a14c1b890d495f708ef23dd118b969c with a reference to #1135.  Consider this tracked there.

File: app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java
Comment (bdferris-v2) [reply app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java]: Updated the TODO in 2251f9435a14c1b890d495f708ef23dd118b969c to reference #1134.

**Comment by maximearmstrong** [comment]
> After #1102 is merged and then the main branch is merged into this PR, then the acceptance tests should pass.

#1102 is now merged, the acceptance tests should pass when the next commit is pushed.

File: app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java
Comment (barbeau) [reply app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java:71]: >Why wouldn't the be organized in a single directory if the GTFS files aren't?

If outputting to temp directly everything gets dumped to that same directory. But if we use the GTFS directory, then it changes depending on where your GTFS zip file is.

>I think it's worth noting that I'd still like to support validation against a URL, so we might still need to find a directory for report output independent of the feed location.

I actually like a combination of 1 and 2 above.

We ask them to pick output, but default to an explicitly named directory off user's home directory. Then save whatever they use for future executions. This aligns with the normal Windows application install experience so it should be familiar to most users, and even non-technical ones.

**Review by maximearmstrong — COMMENTED** [review]
Thanks @bdferris-v2, this is great :)

> I think the module naming is ok. After reading the documentation I think it's obvious, although I'm not sure it would be obvious to someone coming to the project fresh without reading the docs. But I can't think of better naming myself, so I'd say let's go with it :).

After reviewing everything, I agree with this. I think the architecture you provided is the best possible for now. I don't have anything else to add that Sean hasn't already mentioned.

File: app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java
Comment (bdferris-v2) [reply app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/Main.java]: Cool, that sounds good for me.

For now, I've updated the code to use a default output directory ~/GtfsValidator per a2f0567e1f8686e7bf9410a939897b0ddad69272.  This will be used by default for both output reports and logs.

In a follow-up PR where we add in the GUI, we'll allow the user to confirm or change this default and remember their choice for subsequent runs.

**Comment by bdferris-v2** [comment]
Ok, at this point, I think this PR is ready for full review (hopefully we've already done most of the work).  Acknowledged that I've punted a number of implementation details into follow-up work tracked in #1134 and #1135.  However, I think the PR is in a usable state and worth committing before we move onto subsequent iterations.

**Comment by barbeau** [comment]
Could you add the prerequisites and:

```
./gradlew jpackage
```

...that are in the first PR comment in BUILD.md as well?

The prerequisites in the PR comment say "A Java SDK that includes jlink and jpackage". Can you expand on how one would get ahold of such an SDK?

**Comment by bdferris-v2** [comment]
@barbeau done, per 830a9d3ae67a77ec39506f7621329069d5ed43e1

File: app/pkg/build.gradle
Comment (barbeau) [inline app/pkg/build.gradle:90]: I built the exe and it worked!

image

One item - the version reflected here is `1.0`, which doesn't match the actual project version. However, this is a little tricky as the version number is defined by the Git tag in the release process (see RELEASE.md). 

In `main/build.gradle`, it looks like this is passed via an environmental variable - `version = "${System.getenv("versionTag")}"`. Could we use the same process here, and just pass the the version in manually for now when building package from command line (and update the BUILD.md docs accordingly)? 

I believe we could automate the build of the packaged executable with a GitHub Action as well (including grabbing https://github.com/wixtoolset/wix3/releases/tag/wix3112rtm), although I'd suggest we do that in a separate PR.

File: docs/BUILD.md
Comment (barbeau) [inline docs/BUILD.md:42]: I'd suggest removing the `$ ` here so the exact command to execute is what is shown. I've opened PR https://github.com/MobilityData/gtfs-validator/pull/1138 to change this for the other commands in this doc.

File: docs/BUILD.md
Comment (barbeau) [inline docs/BUILD.md:37]: I'd suggest mentioning the WiX requirement here as well.

Also, on two machines (Windows 10 Home and Windows 10 Pro for Workstations) when trying to install the WiX executable from https://github.com/wixtoolset/wix3/releases/tag/wix3112rtm, I got this message:

image

And then I tracked this down for instructions to enable the .NET framework:
https://stackoverflow.com/a/57820594/937715

We may want to include brief instructions here for doing the above workflow.

I was able to install the MSI installer on my Windows 10 Home machine WITHOUT having to enable .NET framework 3.5.1, so it looks like this is only a requirement to build the package, not for end-users installing the package (comments in the above SO post imply that it's required for end users as well, but it wasn't for me).

FYI, for future reference, the packaged MSI is 36.6 MB and the installed application size (as listed by Windows) is 48.9 MB.

**Review by barbeau — COMMENTED** [review]
This is looking good @bdferris-v2! A few minor comments.

File: docs/BUILD.md
Comment (bdferris-v2) [reply docs/BUILD.md:37]: Done in c3588359a1727c9efcc573388911c1e44cd85f6a.

File: app/pkg/build.gradle
Comment (bdferris-v2) [reply app/pkg/build.gradle:90]: This is unfortunately more complicated :)

One immediate issue is with the formatting of version tags in `gtfs-validator`.  Namely, I want to call out that the use of the 'v' prefix is going to be problematic, because many tools (e.g. Java Modules, Mac OS Package Files) expect a purely numeric version number.

I'm also wondering if we can do better than the versionTag env variable.  It seems non-idomatic from a Gradle stanpoint.  It seems like overriding the project.version via a command-line property (-Pversion=1.2.3) seems like a more standard way of doing things?  Indeed, the jpackage command correctly picks up and uses project.version if it's correctly specified (and defaults to 1.0 if not).

Thoughts?

Comment (barbeau) [reply app/pkg/build.gradle:90]: >Namely, I want to call out that the use of the 'v' prefix is going to be problematic, because many tools (e.g. Java Modules, Mac OS Package Files) expect a purely numeric version number.

If this is getting in the way of the other tooling, then 👍  from me for changing it.

>It seems like overriding the project.version via a command-line property (-Pversion=1.2.3) seems like a more standard way of doing things?

👍 , especially if jpackage automatically picks this up. From a quick scan it looks like this would need to change across many of the GitHub Actions .yml files - https://github.com/MobilityData/gtfs-validator/search?q=VERSION_TAG%3Dv

Given that both above changes are more invasive, how about saving them to new PRs so we can get this merged?

Comment (maximearmstrong) [reply app/pkg/build.gradle:90]: > Given that both above changes are more invasive, how about saving them to new PRs so we can get this merged?

I agree. I'm okay with having a`1.0` version for the exe/msi that doesn't match the validator version for now, and then we'll fix that in another PR.

> Namely, I want to call out that the use of the 'v' prefix is going to be problematic, because many tools (e.g. Java Modules, Mac OS Package Files) expect a purely numeric version number.

I agree, we can change it.

Comment (bdferris-v2) [reply app/pkg/build.gradle:90]: I captured this discussion in issue #1139.  I'm going to resolve this for now.

**Comment by bdferris-v2** [comment]
@barbeau @maximearmstrong I think we might be getting close here.  Any additional concerns?

**Review by barbeau — APPROVED** [review]
👍

File: docs/BUILD.md
Comment (maximearmstrong) [inline docs/BUILD.md:32]: I think that since the Mac OS application doesn't work at the moment (I assume the same for Linux), we could include a note saying that the installable application can be built on any OS, but only works on Windows at the moment. It could be here or in the `README.md`.

**Review by maximearmstrong — CHANGES_REQUESTED** [review]
I agree, we are very close. One thing, I think we need a section in the `README.md` on how to use the installable application. I also added a comment related to that in my review.

File: docs/BUILD.md
Comment (bdferris-v2) [reply docs/BUILD.md:35]: I added a note in 5e02c89531fd662b202bdb716f18d4772c533bad

**Comment by barbeau** [comment]
>One thing, I think we need a section in the README.md on how to use the installable application. I also added a comment related to that in my review.

I almost commented saying something similar, but after more thought I would wait on this, because it's going to change significantly when the UI is added in an upcoming PR.  I think the existing note saying that it's under active development in the build section is probably enough for now as it's not ready to be user-facing yet.

**Comment by maximearmstrong** [comment]
> > One thing, I think we need a section in the README.md on how to use the installable application. I also added a comment related to that in my review.
> 
> I almost commented saying something similar, but after more thought I would wait on this, because it's going to change significantly when the UI is added in an upcoming PR. I think the existing note saying that it's under active development in the build section is probably enough for now as it's not ready to be user-facing yet.

I agree. I was about to say the same now that the note have been added.

**Review by maximearmstrong — APPROVED** [review]

**Comment by maximearmstrong** [comment]
@bdferris-v2 I approved this PR. Thank you for this contribution! 🚀

**Comment by maximearmstrong** [comment]
@bdferris-v2 It seems that we are only missing the signature of the license/cla to be able to merge.

**Comment by bdferris-v2** [comment]
@maximearmstrong I did sign the CLA, like two weeks ago :)

When I go to https://cla-assistant.io/MobilityData/gtfs-validator, I see the following:

Screen Shot 2022-05-10 at 3 23 33 PM

Is there something else I need to do?  Anything you want to check on your end?

**Comment by barbeau** [comment]
@maximearmstrong I think my permissions on the MobilityData org changed at some point and I can no longer see the list of signed CLAs for MobilityData:
image

I also can't see the current hooks/3rd party tools integrations on this repo settings either. IIRC I'm the one that set this up originally.

Could you try going to https://cla-assistant.io/ and see if you can see @bdferris-v2's CLA there?

I'm wondering if something is wrong with the CLA bot, although it looks like a 3rd party contributor was recognized as having signed the CLA by the bot in Feb:
https://github.com/MobilityData/gtfs-validator/pull/1090

I'm not sure if something re: permissions would have changed since then that could have broken it. Maybe related to https://github.com/MobilityData/gtfs-realtime-validator/issues/150?

**Comment by maximearmstrong** [comment]
@bdferris-v2 Oh, it seems the issue is that your other account @bdferris is also a committer on the project. Since you got locked out of that account and signed the CLA with your new account, I think we can merge without any problem.

Capture d’écran, le 2022-05-10 à 18 30 07

**Comment by bdferris-v2** [comment]
Yeah, I'm sure that's going to cause continued confusion.  The original account is still associated with my primary email, so it tends to get picked up by default despite my best efforts.

Either way, I'm going to go ahead and merge.  Thanks!

**Comment by maximearmstrong** [comment]
> Yeah, I'm sure that's going to cause continued confusion. The original account is still associated with my primary email, so it tends to get picked up by default despite my best efforts.

No worries. I have reconfigured the CLA assistant using my account, so we'll see. You may need to sign it again on your next PR. Thanks again for your contribution!

**Pull request merged**
